### DISCUSSION TITLE
Macros! TW blocks that compile to scratchblocks

### DISCUSSION THREAD

**Discussion by OceanIsEndless** [TurboWarp/extensions#1592] [Ideas]
# Macros!

Hello, I’m new here, and I am curious as to if there is (or could soon be) a way to have **macros** in extensions!

## TL;DR

> [what this person is suggesting is custom extension blocks that compile to _vanilla scratch blocks_](https://github.com/TurboWarp/extensions/discussions/1592#discussioncomment-9976980)

Plz add the ability to put special blocks which will turn into scripts at compile time into custom extensions. Thanks! **`THE END`**

## Introduction

Basically, say I wanted to create the following blocks for a custom extension:

![<true> imaginary boolean block](https://github.com/TurboWarp/extensions/assets/154556928/365680f4-da59-4941-abe7-e6a83e5e0d83)

![<false> imaginary boolean block](https://github.com/TurboWarp/extensions/assets/154556928/3a24f1c4-7b84-4e5a-bd57-1d50623dc718)

I know that these are already featured in existing **TurboWarp** extensions. However, I want to be able to use these custom _TurboWarp_ blocks in _Scratch_. Fortunately, there are pretty easy workarounds for these blocks:

![<not <>> workaround for <true>](https://github.com/TurboWarp/extensions/assets/154556928/950c76fb-5a23-4626-b67b-e4d50dbb52d7)

![<not <<not <>>> workaround for <false>](https://github.com/TurboWarp/extensions/assets/154556928/abed4a82-0f7b-410b-b67e-aadb71f384c3)

However, nobody wants to have to make their code into an unreadable mess such as this:

![Unreadable code using the workarounds for the <true> and <false> blocks](https://github.com/TurboWarp/extensions/assets/154556928/95761ffa-46b4-43d2-957f-f1b1b3ccb47d)

<details><summary>Rant</summary>
<p>
<small><i>Yes, I know strings exist. However, this is just a very simple example here, and there are plenty more. Additionally, writing the word "true" actually behaves differently than the actual boolean true, and writing the number 1 is still kind of unreadable. But I did not come here to argue with you about the usefulness of this example anyway. >:(</i></small>
</p>
</details> 

**What I’m proposing is,** there should be a way to have **TurboWarp** blocks be _automatically converted_ to their **Scratch** counterparts at _compile time_.

## What a Macro Is

Yes, we have "My Blocks", but they can be limiting in functionality, and _macros_ are a whole different concept. You see, a _custom block_ is a script which executes at _runtime_, its code being **referenced** by other scripts. A _macro_, on the other hand, is a script which runs at _compile time_. It produces blocks which are **injected** into the script wherever the macro is being used. The macro is not included in the script, but rather the blocks that the macro produces are. Macros, at their most basic, are simply a shorthand way of writing blocks or scripts, but with some custom logic, **they can be _much_ more than that.** Macros are really just extension blocks that turn into ordinary scratch blocks, which can then be run by TurboWarp, and better yet Scratch.

> ### Edit:
>
> So I realized that the way I explained it here was kind of unclear. So to clarify- macros are "custom extension blocks that compile to _vanilla scratch blocks_", THANK YOU @CST1229 for that [incredibly clear statement](https://github.com/TurboWarp/extensions/discussions/1592#discussioncomment-9976980) which pretty much sums up this entire topic lol.

## Uses

Macros have a **GREAT** potential for projects. Some use cases include:

- **Readability!** You can **look at it** and **actually know** what it does! <small><i>Technically only in TurboWarp, as the compiled Scratch output would be using the workarounds, however it would still provide for a greatly improved programming experience!</i></small>
- **Pre-Processing!** **Process** block **code** ahead of time **using JavaScript** in whatever way that best suits your project’s needs!
- **Evaluate things ahead of time!** Do it so that your project doesn’t have to! _Process things that need processing!_
- **Storage!** Store your files elsewhere (on a server, rather than hardcoding them) and fetch them at compile time! _You may have to handle cases where you can’t access the file though._
- **DRY!** Don’t Repeat Yourself! _Macros technically repeat, but at least they’re much easier to write!_
- **Convenience!** Speed up your workflow with **macros**! _It makes script-making as easy as dragging a block out of the palette!_

Yes, I know that this sounds largely like a convenience feature so far, but I encourage you to read to the end, as macros can actually be quite a powerful tool.

#### Simple Example

 Here’s one of the _simpler_ examples:

![A scratchblocks example showing off an imaginary macro, multiply [variable v] by (amount)](https://github.com/TurboWarp/extensions/assets/154556928/38b9daeb-a894-4c50-a999-53664c349708)

<details><summary>scratchblocks</summary>
<pre>
<code>
multiply [my variable v] by (3) :: variables // TurboWarp Macro<br>
<br>
set [my variable v] to ((my variable) * (3)) // Scratch Workaround<br>
<br>
when I receive [5x BOOST! v] // Example Script<br>
say [WEEEEEEEEEEEEEEEEEE]<br>
multiply [speed v] by (5) :: variables // TurboWarp Macro<br>
<br>
when I receive [5x BOOST! v] // Compiled Output<br>
say [WEEEEEEEEEEEEEEEEEE]<br>
set [speed v] to ((speed) * (5)) // Scratch Workaround
</code>
</pre>
</details>

The macro shown off above would be a shorthand way of multiplying a variable by a certain amount. Doing this with My Blocks would be no easy feat, as you cannot programmatically determine which variable to set in ordinary Scratch. However, since **macros** run at _compile time_ rather than at _runtime_, using some extra JavaScript, the macro can magically transform itself into ordinary block code based off of its arguments before it gets sent to the scratch-vm. Do you see what I’m getting at now??? :P

#### Complex Example

But macros can do **WAY** more than just that. Here’s some more _complex_ examples:

`<insert some more complex examples here because I’m too lazy to put them here… sorry! ZzzZZZzzZz… I’m tired, OK?>`

For real though, the possibilities with this are great. For example, one could program a reporter macro that fetches a file from a server and converts it into a format that your project can use at runtime. Or, one could create a macro for checking the type of a value. OR, you could even make a macro of macros. **WHO KNOWS?!?!?!?**

_I’ve spent a lot of time writing this, OK? Gimme a break…_

There are a million other examples that I could give you, but I think that you should be getting the point by now. :>

### Insertions

It would be cool if macros could also add statements ***around*** them, like so:

![A <confirm ()> boolean macro which inserts an ask () and wait stack block before it in the script for the Scratch workaround](https://github.com/TurboWarp/extensions/assets/154556928/2b8c51fb-980c-4099-96f1-7ab15a80190a)

This macro asks the user for confirmation as a boolean. To ask the user, it inserts an `ask () and wait` block _before_ the statement that the macro was used in. This allows for more complex reporter macros with multiple steps to them!

## Implementation

I think that **macros** could be added in one or more of the following ways:

1. **Additional options** specifically for _making macros™️_ inside of custom TurboWarp extensions
2. **Packager settings** for macros (no clue how that would work or what it would do, but it’d be cool anyhow)
3. **"My Macros" category**/special extension similar to My Blocks, but for macros (might be asking a little too much here, but we’ll see :)

I think that the **first** and **third** options, _specifically_ the **first** (adding additional options for macros to custom extensions) would be the most viable and practical. A macro would be defined using JavaScript in a similar way to an ordinary block, the main difference being its logic would be called at _compile time_, and it would have to somehow be able to produce the "compiled" code for itself _(or perhaps even modify the code around it for more complex macros)_. This _could_ require some major reworking, however I’m sure that if anyone can do it, it’s you. :D

There may have to be a separate option when exporting a project from TurboWarp so that one can choose whether or not to download the file with _the macros included in it_ or with _the macros compiled_. It’s clear that a lot would have to happen to get macros added into TurboWarp, but I hope that we can iron all that stuff out in order to create **an amazing, much-needed** _(well, maybe_ you _didn’t)_ **original** _(don’t quote me on that)_ **feature that will greatly change the idea of block-programming as we know it!**

## Notes

- **When I say "compiled" here, I don’t mean the compiled JavaScript code that TurboWarp produces,** but rather the scratchblocks that the macro would produce. When I say "compile time", I mean either before a project is executed or before a project is saved as a file.
- **Macros are by no means a replacement for custom blocks, nor are custom blocks a workaround for macros.** Both have their own incredible uses and can coexist peacefully.
- **Macros would probably be reserved for unsandboxed extensions.** Being able to modify a program’s scripts may be a bit too powerful for the sandbox. _Although I guess if done right it could also be available to sandboxed extensions... I don’t know man XD_
- **Macros can have multiple steps to them.** I don’t want this to just be a block switcheroo; I want this thing to create whole **SCRIPTS** using just **ONE** BLOCK!!! _No clue how to make any of this work, I hope you all can figure that out for me. I just had an idea, and I’m leaving the adding-it-in part up to you._
- **I haven’t made an issue for this yet.** I was kind of hoping that somebody else could do that for me, as this is just an idea I had and I’m not _exactly_ sure how it would be incorporated into TurboWarp yet. I just want to get the discussion rolling first so that we can figure out how exactly macros would really work before we go about implementing them. Hopefully you understand. :)
- **Sorry for any mistakes!** If you couldn’t tell, this thing took me an eternity to write. I may have said something incorrectly or not explained something well enough. Feel free to ask questions! _Although I can’t guarantee that I’ll answer XD. I’m new to GitHub and posting on it, so hopefully I can figure all that stuff out. I’ve been using TurboWarp for a long long while now, I just haven’t bothered saying anything or contributing really until this idea came to my brain. I hope you all find this a good idea too!_
- **Macros are really cool.** I hope you've learned that by the end of this post. ;P

## Conclusion

Adding macros to TurboWarp would create new and exciting possibilities for Scratch projects, and would open the door to a whole new level of complexity in what started as a simple block-based programming language. I am exhausted after writing all of this, so I’ll leave the rest of the discussion to you guys. I am excited to hear what everybody has to say about this, and thanks for reading. :]

…

**Thank you for coming to my TED Talk.**

**Comment by namelessisbackbutnottodobadthingsiswear** [comment]
I've though about similar things in the past, however the Scratch Team likely wouldn't take to fondly to it. Specifically, there's a clause in the Terms of Use ([section 4.4](https://scratch.mit.edu/terms_of_use)) that prohibits uploading projects that cannot be run on Scratch. It's different for projects that simply aren't powerful enough for Scratch, given with enough time and sacrifice they _can_ run on the website, however designing projects specifically to run in TurboWarp and not Scratch would definitely be against TOS.
If your project requires TurboWarp extensions to run, the best course of action is to share it somewhere other than Scratch. Otherwise, you'd be putting your account (and by extension TurboWarp) against the TOS.

**Reply by CST1229** [reply]
i think what this person is suggesting is custom extension blocks that compile to _vanilla scratch blocks_, which someone already did: https://github.com/ZXMushroom63/scratch-gui

**Reply by OceanIsEndless** [reply]
> I've though about similar things in the past, however the Scratch Team likely wouldn't take to fondly to it. Specifically, there's a clause in the Terms of Use ([section 4.4](https://scratch.mit.edu/terms_of_use)) that prohibits uploading projects that cannot be run on Scratch. It's different for projects that simply aren't powerful enough for Scratch, given with enough time and sacrifice they _can_ run on the website, however designing projects specifically to run in TurboWarp and not Scratch would definitely be against TOS. If your project requires TurboWarp extensions to run, the best course of action is to share it somewhere other than Scratch. Otherwise, you'd be putting your account (and by extension TurboWarp) against the TOS.

Oof, I guess I needed to be a bit more clear in my post… XD This is kind of a weird idea, and I probably should have explained it better. I’ll try to rephrase it a little.

Firstly, macros are **NOT** blocks. They are really special functions that _turn into_ blocks. They are not stored in the project that is uploaded to Scratch. They are blocks that produce scripts. I know it’s kind of confusing, so here’s an example:

![<true> imaginary block](https://github.com/TurboWarp/extensions/assets/154556928/92c5186a-7df3-41d8-93c3-afd4f8ccb1df)

I want to be able to use this block in Scratch. But for obvious reasons, I can’t do that _directly_. However, I _can_ work around it like so:

![<not <>> workaround for the <true> imaginary block](https://github.com/TurboWarp/extensions/assets/154556928/243b4fd9-61d4-49ec-b99b-29707d96be13)

"Macros" are basically blocks which turn into their workarounds. So instead of it being a _block_ in of _itself_, it turns _itself_ into a _block_. To put it simply, whenever I use a "macro" in a project, if TurboWarp tries to **compile** or **export** it (there would be an option to download it with the macros included or not) then the special block/macro is called. The macro outputs a script which is then substituted where the block once was.

So if I used the macro:

![<true> imaginary block](https://github.com/TurboWarp/extensions/assets/154556928/92c5186a-7df3-41d8-93c3-afd4f8ccb1df)

And TurboWarp tries to **compile it** or **export it** to Scratch, the macro would turn itself into:

![<not <>> workaround for the <true> imaginary block](https://github.com/TurboWarp/extensions/assets/154556928/243b4fd9-61d4-49ec-b99b-29707d96be13)

Keep in mind that the script in the editor remains the same. It’s just that whenever the macro is used by TurboWarp (either **compiled** or **downloaded** for Scratch), it will simply be interpreted as if it were the script that was returned by the macro (the workaround) rather than the macro itself. So when I upload it to Scratch, instead of the macro being there, it is changed into the vanilla Scratch script instead.

I hope that clears it up a little bit…? Tell me if it needs further clarifying lol. In short, macros are special blocks which turn into scripts wherever they are used. Again, feel free to ask if you still don’t get it. I just want to clear up any confusion I can. :)

**Reply by OceanIsEndless** [reply]
> i think what this person is suggesting is custom extension blocks that compile to _vanilla scratch blocks_, which someone already did: https://github.com/ZXMushroom63/scratch-gui

That sounds about right. (Hopefully we can still have it in TurboWarp though?)

**Comment by snowboyz0825** [comment]
I feel like a much better way to describe this would be that they want to add blocks that make building projects easier, such as "Multiply x by 5," and appear that way in the editor, but when saved it saves as "Set x to (x)(5)." Also potentially have a few ways to "Convert" certain extension blocks to a scratch block. 

As much as I've wanted this a TON, it's really something you just have to do manually. It's impossible to convert most extension code into scratch code without changing it entirely, and that's something up to the user's discretion. Maybe one day I can understand the SB3 jsons enough to code one for myself, but that doesn't guarantee that it will work for others, and I'm 99% sure that before anything gets added to turbowarp has to work for everyone, and no matter how customizable you make it, it's too much work to make it work for everyone.
